docs/source/reference/database.md

```diff
@@ -1,29 +1,120 @@
+This is the **state** of the Hub.
 
-## Default SQLite database
+## Why does JupyterHub have a database?
 
-The default database for JupyterHub is a [SQLite](https://sqlite.org) database.
-We have chosen SQLite as JupyterHub's default for its lightweight simplicity
-in certain uses such as testing, small deployments and workshops.
+JupyterHub is a **stateful** application (more on that 'state' later).
+Updating JupyterHub's configuration or upgrading the version of JupyterHub requires restarting the JupyterHub process to apply the changes.
+We want to minimize the disruption caused by restarting the Hub process, so it can be a mundane, frequent, routine activity.
+Storing state information outside the process for later retrieval is necessary for this, and one of the main thing databases are for.
+
+A lot of the operations in JupyterHub are also **relationships**, which is exactly what SQL databases are great at.
+For example:
+
+- Given an API token, what user is making the request?
+- Which users who don't have running servers?
+- Which servers belong to user X?
+- Which users have not been active in the last 24 hours?
+
+Finally, a database allows us to have more information stored without needing it all loaded in memory,
+e.g. supporting a large number (several thousands) of inactive users.
+
+## What's in the database?
+
+The short answer of what's in the JupyterHub database is "everything."
+JupyterHub's **state** lives in the database.
+That is, everything JupyterHub needs to be aware of to function that _doesn't_ come from the configuration files, such as
+
+- users, roles, role assignments
+- state, urls of running servers
+- Hashed API tokens
+- Short-lived state related to OAuth flow
+- Timestamps for when users, tokens, and servers were last used
+
+### What's _not_ in the database
+
+Not _quite_ all of JupyterHub's state is in the database.
+This mostly involves transient state, such as the 'pending' transitions of Spawners (starting, stopping, etc.).
+Anything not in the database must be reconstructed on Hub restart, and the only sources of information to do that are the database and JupyterHub configuration file(s).
+
+## How does JupyterHub use the database?
+
+JupyterHub makes some _unusual_ choices in how it connects to the database.
+These choices represent trade-offs favoring single-process simplicity and performance at the expense of horizontal scalability (multiple Hub instances).
+
+We often say that the Hub 'owns' the database.
+This ownership means that we assume the Hub is the only process that will talk to the database.
+This assumption enables us to make several caching optimizations that dramatically improve JupyterHub's performance (i.e. data written recently to the database can be read from memory instead of fetched again from the database) that would not work if multiple processes could be interacting with the database at the same time.
```

Comment:
No action point or issue seems relevant to create as a followup, just felt like sharing this brief thought!

This got me thinking that maybe a way to incrementally work towards supporting multiple replicas of a JupyterHub is to make such information requests flip from being read in memory to being read from the database or similar.

Reply:
Yes, I've outlined the steps involved [here](https://github.com/jupyterhub/jupyterhub/issues/1932#issuecomment-658031232) and [here](https://figshare.com/articles/poster/The_cloud_is_the_limit_scaling_limits_of_JupyterHub/14937954), but it's roughly:

1. take away direct db access from Authenticator and Spawner (already deprecated, _almost_ never used)
2. remove the in-memory Users dict
3. open a db session per-request instead of using a shared session

These are all that's needed for the db session. It will have a pretty large performance cost for idle-culling, because listing all 'active' users is an iteration over a Python dict of already-constructed objects, rather than recreating a bunch of ORM objects from a db query, which it will _have_ to be going forward. That's orders of magnitude faster. _But_ it's also a scaling issue because all 'active' users must fit in memory.

The main challenge is that User objects are ORM wrappers and Spawners have access to these, and Spawners are long-lived objects. Providing them with a read-only static snapshot of the User data ought to be fine for almost everyone, but it will need to be a breaking API change. Cases that will be hard are e.g. checking user group membership, which some folks do now in hooks. That will mean not only fetching the user, but fetching all their _relationships_ that we want to support Spawners _possibly_ accessing, rather than what we have now, which is allowing them to request anything as-needed. It's probably better in the long run! But it will be a hard transition. And if we don't enable multiple replicas in the same release, it will be a purely _worse_ JupyterHub as a result.

That's largely why I haven't started work on it - most of the incremental changes, while possible, are _detrimental_ to JupyterHub until they reach the final goal of enabling replicas. So I don't really want to start it until I have the support needed to finish it.

Reply:
I find this kind of information about _current_ plans about the project _future_ very useful for creating a mental picture about the general direction of where the work should be put into and how to make sustainable technical choices and that don't harm the general plan. But currently, as it is, as GitHub issues, this info is pretty hard to find when searching for them intentionally unless, you already know some keywords and have some prior context.

So a question popped into my head when reading this, that I feel was discussed before (?) is "Where should we store such project plan in a place that is findable, iterable and easy to update when things change?" Is it discourse, the documentation itself, or GitHub issues / projects?

(cc @sgibson91 as having such place, I believe might be of help for bringing in, getting feedback from and keeping new contributors in the community)

Reply:
@GeorgianaElena There's an out of date roadmap https://jupyterhub.readthedocs.io/en/stable/contributing/roadmap.html 
Might be worth a team-compass discussion for updating or replacing it?

Reply:
Yeah, roadmaps are a great tool for this but they need regular(ish, depending on how long the roadmap is intended to last) tending/updating. So then next after the team-compass discussion on updating it, we'd need another discussion about how to organise regular updates to the roadmap. It can be a time-consuming process that can be difficult to coordinate for, but very beneficial for the project and contributors.

Reply:
I wonder if GitHub projects might be a better way to represent some of these things?

I do think the hub team is due for a 'roadmap workshop' to review/reset/communicate medium and long-term goals across the project(s).

Reply:
Oooh, I like both of these ideas!

Reply:
I'm in favour of not using a static RTD page for the roadmap.... that's why I suggested a new team-compass issue, instead of sidetracking this PR 😄